Configure TLS for private v2 registry mirrors.

RichardScothern opened the pull request:
Signed-off-by: Richard Scothern richard.scothern@gmail.com

Closes #14823 

cc @tiborvass @dmcgowan

rscothern committed:
Configure TLS for private registry mirrors.

If a registry mirror is using TLS, ensure that certs for it
are picked up from /etc/docker/certs.d

Signed-off-by: Richard Scothern <richard.scothern@gmail.com>

tiborvass commented on registry/service.go:
```diff
@@ -161,19 +162,35 @@ func (s *Service) TlsConfig(hostname string) (*tls.Config, error) {
+	if err != nil {
+		return &tls.Config{}, err
+	}
+	tlsConfig, err := s.TlsConfig(mirrorUrl.Host)
```
you can simply `return s.TlsConfig(...)`

calavera replied:
agreed with this.

calavera commented on registry/service.go:
```diff
@@ -161,19 +162,35 @@ func (s *Service) TlsConfig(hostname string) (*tls.Config, error) {
+func (s *Service) tlsConfigForMirror(mirror string) (*tls.Config, error) {
+	mirrorUrl, err := url.Parse(mirror)
+	if err != nil {
+		return &tls.Config{}, err
```
this can be `return nil, err`

calavera commented on registry/service.go:
```diff
@@ -161,19 +162,35 @@ func (s *Service) TlsConfig(hostname string) (*tls.Config, error) {
+	}
+	tlsConfig, err := s.TlsConfig(mirrorUrl.Host)
+	if err != nil {
+		return &tls.Config{}, err
```
this can be `return nil, err`

stevvooe commented:
LGTM after fixing the error returns.

tiborvass commented:
I wonder what happens with v1 mirrors

dmcgowan commented:
Can you clarify what the original behavior was that was causing problems?

tiborvass commented:
@dmcgowan the original problem was that the tls certs were not set up for v2 mirrors.

calavera commented:
without calling the `s.TlsConfig` function for each mirror you could try to use certificates for other host incorrectly.

dmcgowan commented:
@tiborvass to clarify my understanding, the issue is that it was not using certificates from `/etc/docker/cert.d` for mirrors, which could include needed client certificates. It should not have been using the wrong certificates since it would have defaulted to the base configuration.

dmcgowan commented:
> I wonder what happens with v1 mirrors

@tiborvass To your comment above the section, I don't think v1 mirrors should even be included in that list. V1 mirrors do not mirror the index and those endpoints should only be indexes.

tiborvass commented:
@RichardScothern feel free to remove v1 mirrors in that LookupEndpoints list.

rscothern committed:
Remove v1 registry mirror configuration from LookupEndpoints.

V1 mirrors do not mirror the index and those endpoints should
only be indexes.

Signed-off-by: Richard Scothern <richard.scothern@gmail.com>

dmcgowan commented:
LGTM

tiborvass commented:
Let's make it clearer in the commit message and PR description that this is fixing *private* v2 mirrors and public v2 mirrors were working just fine.

tiborvass commented:
LGTM
